### PR TITLE
Updated API

### PR DESCRIPTION
**Pull request by anantn**
This merge is for moving the new mozApps code for both the HTML5 & add-on versions.

### CODE REVIEW

File: addons/jetpack/lib/api.js [hunk 1]
```diff
@@ -121,6 +168,42 @@ FFRepoImpl.prototype = {
...
 
+  // Implement getSelf, getInstalled to use repo.js;
+  // *but* wrap them in an appobject (which exposes launch & uninstall)
+  // Can't do this for getAll since mgmt functions are implement using
+  // worker.emit and pageMods. The equivalent logic for the following two
+  // functions for getAll lives in mgmtapi.js
+  getSelf: function(origin) {
+    let pendingGetSelf = new pendingOperation("getSelf");
+    Repo.getSelf(origin, function(app) {
+      // FIXME: what's the error case?
```
Comment (mixedpuppy) [inline addons/jetpack/lib/api.js]: is a bug for missing error case necessary?

File: addons/jetpack/lib/main.js [hunk 2]
```diff
@@ -545,7 +558,8 @@ function startup(getUrlCB) { /* Initialize simple storage */
   Services.obs.notifyObservers(tmp.FFRepoImplService, "openwebapps-startup-complete", "");
 
   // initialize the injector if we are <fx9
-  require("./injector").init();
+  // we don't need to support fx<9 anymore, TODO: remove injector.js
```
Comment (mixedpuppy) [inline addons/jetpack/lib/main.js]: remove these comments then, and delete the injector code

File: addons/jetpack/lib/api.js [hunk 3]
```diff
@@ -181,18 +264,26 @@ FFRepoImpl.prototype = {
 
   },
 
+  // FIXME: return pendingAppInstall object, remove onsuccess, onerror
```
Comment (mixedpuppy) [inline addons/jetpack/lib/api.js]: why?  what's the explanation, what's the side effect, where's the bug#

File: addons/jetpack/lib/main.js [hunk 4]
```diff
@@ -227,25 +225,40 @@ MozAppsAPI.prototype = {
...
-          url: origin, install_data: data,
-          onsuccess: onsuccess, onerror: onerror
+          url: path, installData: data
         };
-        repo.install(aWindow.location, args, recentWindow);
+
+        // If install is triggered from 127.0.0.1, don't show doorhanger
+        // FIXME: how can we implement this without giving all of 127.0.0.1
+        // the privilege to bypass the doorhanger?
+        let domain = "http://127.0.0.1";
```
Comment (mixedpuppy) [inline addons/jetpack/lib/main.js:237]: bug#?  what is the possible risk associated with this?

File: addons/jetpack/lib/main.js (hunk 2 above)
Comment (anantn) [inline addons/jetpack/lib/main.js]: Fixed: https://github.com/mozilla/openwebapps/commit/861955a835c37700ccd773a534d0ff16362258fa

File: addons/jetpack/lib/api.js (hunk 3 above)
Comment (anantn) [inline addons/jetpack/lib/api.js]: Oops, that was just a mental note to remind me to do it; the code does infact return a pendingAppInstall object and does not use onsuccess/onerror (to move it in line with the new API interface) - I had just forgotten to remove the comment after making the changes. Fixed: https://github.com/mozilla/openwebapps/commit/e6249ab70ce982f2d1ccb4e23379bb688a94b674

File: addons/jetpack/lib/api.js (hunk 1 above)
Comment (anantn) [inline addons/jetpack/lib/api.js]: Repo.getSelf doesn't expose an error callback; I think error conditions are handled by an empty argument to the success callback. I've updated the comment to reflect that.

File: addons/jetpack/lib/main.js (hunk 4 above)
Comment (anantn) [inline addons/jetpack/lib/main.js:237]: https://bugzilla.mozilla.org/show_bug.cgi?id=727255

I think the risk is minimal, as malicious script would have to be served off 127.0.0.1 in order to exploit this (by installing apps without the user's permission). If you're able to serve content off 127.0.0.1 there's probably worse things you could do :)

The bigger problem is to be able to test the add-on in an automated fashion without adding workarounds like this.

**Pull request merged**